### PR TITLE
Design updates for #2015 numeric type literal syntax

### LINKED ISSUE #2159: Need design updates for #2015 numeric type literal syntax
We need updates to docs/design corresponding to proposal #2015.

### PR DESCRIPTION
**Pull request by avicorp**
Edit numeric literal design for the literal type proposal and add reference where #2015 was mentioned.

Closes #2159

### CODE REVIEW

**Comment by chandlerc** [comment]
Thanks for the design updates! Adding @josh11b to help with review here as he helped review the original proposal.

File: docs/design/README.md [hunk 1]
```diff
@@ -477,8 +476,6 @@ may be limited to integers of at most 128 bits due to LLVM limitations.
 >     [#861: Naming conventions](https://github.com/carbon-language/carbon-lang/pull/861)
 > -   Proposal
 >     [#1083: Arithmetic expressions](https://github.com/carbon-language/carbon-lang/pull/1083)
-> -   Proposal
->     [#2015: Numeric type literal syntax](https://github.com/carbon-language/carbon-lang/pull/2015)
```
Comment (josh11b) [inline docs/design/README.md]: Seems better to keep this -- it still seems relevant to readers of this section who want to learn more.

File: docs/design/README.md [hunk 2]
```diff
@@ -542,8 +540,6 @@ number.
...
```
Comment (josh11b) [inline docs/design/README.md]: I would keep this too.

File: docs/design/README.md [hunk 3]
```diff
@@ -469,6 +467,7 @@ may be limited to integers of at most 128 bits due to LLVM limitations.
 
 > References:
 >
+> -   [Numeric Literal Semantics](numeric_literals.md)
```
Comment (josh11b) [inline docs/design/README.md]: This should link to a new document, not one about numeric literals.

File: docs/design/README.md [hunk 4]
```diff
@@ -534,6 +531,7 @@ number.
...
```
Comment (josh11b) [inline docs/design/README.md]: This should also be a link to the new document that this PR should be adding.

File: docs/design/numeric_literals.md [hunk 5]
```diff
@@ -13,10 +13,8 @@ SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 ## Table of contents
```
Comment (josh11b) [inline docs/design/numeric_literals.md]: I don't think you should be changing this document. This document is talking about numeric literals, like `0`, `-7`, and `3.14`, and we want to keep this content. Numeric type literals are a different thing, and would be covered by a different document.

**Review by josh11b — CHANGES_REQUESTED** [review]

File: docs/design/numeric_literals.md (hunk 5 above)
Comment (avicorp) [reply docs/design/numeric_literals.md]: I see. I will create new md document for numeric type literals.

**Review by josh11b — COMMENTED** [review]
Please follow the instructions at https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md to set up `pre-commit` and `prettier` for auto-formatting.

**Comment by josh11b** [comment]
I talked with @zygoloid about how we should organize the docs, and he had the idea, which I like, that:
- eventually the `docs/design/expressions/` section describes all kinds of expressions, which would include numeric type literals
- it would be reasonable to have a single document that covers all kinds of literals, especially if we don't have much to say about them, and only break out specific kinds of literals to their own documents if there are enough details to warrant that  (more than 3 pages of text, say)

The path forward would be to move this content into `docs/design/expressions/literals.md` and link to it from `docs/design/expressions/README.md`; if `literals.md` gets too big we can think about splitting it up then.

**Comment by avicorp** [comment]
@josh11b, how can we continue with the PR?

File: docs/design/expressions/literals.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
+# Numeric type literal expression
```
Comment (josh11b) [inline docs/design/expressions/literals.md]: This document is intended to cover all literal expressions, even if it for now only covers numeric type literals.

```suggestion
# Literal expressions
```

File: docs/design/expressions/literals.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+## Table of contents
+
+-   [Overview](#overview)
+    -   [Usage](#usage)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
```
Comment (josh11b) [inline docs/design/expressions/literals.md:21]:
```suggestion
## Numeric type literals
```

File: docs/design/expressions/README.md [hunk 8]
```diff
@@ -298,6 +299,15 @@ fn Run(args: Span(StringView)) {
...
 `if` expressions are analogous to `?:` ternary expressions in C and C++.
 
+## Numeric type literals expressions
+
+Carbon's syntax provides a simple way to represent different types of integers
+and floating-point numbers. Each type is identified with a keyword-like syntax,
+prefixed with either `i`, `u`, or `f` followed by a multiple of 8, representing
+the size of the data type.
+
+The types referred to as [Numeric type literals expression](literals.md).
```
Comment (zygoloid) [inline docs/design/expressions/README.md]:
```suggestion
The types are referred to as [numeric type literals expressions](literals.md).
```

File: docs/design/expressions/literals.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+fn Main() -> i32 {
+  return Sum(4, 2);
+}
+```
+
+In the above example, `Sum` has parameters `x` and `y`, each of which is typed
+as a 32-bit two's complement signed integer. `Main` then returns the output of
+`Sum` as a 32-bit two's complement signed integer.
+
+## Alternatives considered
```
Comment (josh11b) [inline docs/design/expressions/literals.md]: Once we add other kinds of literals, these will only be the alternatives for numeric type literals.
```suggestion
### Alternatives considered
```

File: docs/design/expressions/literals.md (hunk 6 above)
Comment (zygoloid) [inline docs/design/expressions/literals.md]:
```suggestion
# Numeric type literal expressions
```

File: docs/design/expressions/literals.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+-   [Overview](#overview)
+    -   [Usage](#usage)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+Carbon's introduce a simple keyword-like syntax of `iN`, `uN`, and `fN` for
```
Comment (zygoloid) [inline docs/design/expressions/literals.md]:
```suggestion
Carbon has a simple keyword-like syntax of `iN`, `uN`, and `fN` for
```

File: docs/design/expressions/literals.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+two's complement integers, unsigned integers, and
+[IEEE-754](https://en.wikipedia.org/wiki/IEEE_754) floating-point numbers,
+respectively. Where `N` can be a positive multiple of 8, including the common
```
Comment (zygoloid) [inline docs/design/expressions/literals.md]:
```suggestion
respectively. Here, `N` can be a positive multiple of 8, including the common
```

File: docs/design/expressions/literals.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+-   [Alternatives considered](/proposals/p2015.md#alternatives-considered)
+
+## References
+
+-   [Rationale](/proposals/p2015.md#rationale)
+-   Proposal
+    [#2015: Numeric type literal syntax](https://github.com/carbon-language/carbon-lang/pull/2015)
+-   Issue
+    [#1998: Make proposal for numeric type literal syntax](https://github.com/carbon-language/carbon-lang/issues/1998#issuecomment-1212644291)
```
Comment (josh11b) [inline docs/design/expressions/literals.md]: This was the motivation for the current PR, but is not something that would add value to a reader of this file.

```suggestion
```

File: docs/design/expressions/literals.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+as a 32-bit two's complement signed integer. `Main` then returns the output of
+`Sum` as a 32-bit two's complement signed integer.
+
+## Alternatives considered
+
+-   [Alternatives considered](/proposals/p2015.md#alternatives-considered)
+
+## References
+
+-   [Rationale](/proposals/p2015.md#rationale)
```
Comment (josh11b) [inline docs/design/expressions/literals.md]: Readers can find that by following the link to the proposal on the next line. Better to include the leads question that had discussion about the options.
```suggestion
-   Issue
    [#543: pick names for fixed-size integer types](https://github.com/carbon-language/carbon-lang/issues/543)
```

File: docs/design/expressions/literals.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+}
+```
+
+In the above example, `Sum` has parameters `x` and `y`, each of which is typed
+as a 32-bit two's complement signed integer. `Main` then returns the output of
+`Sum` as a 32-bit two's complement signed integer.
+
+## Alternatives considered
+
+-   [Alternatives considered](/proposals/p2015.md#alternatives-considered)
```
Comment (josh11b) [inline docs/design/expressions/literals.md]: Better to list the individual alternatives here:
```suggestion
-   [C++ LP64 convention](/proposals/p2015.md#c-lp64-convention)
-   [Type name with length suffix](/proposals/p2015.md#type-name-with-length-suffix)
-   [Uppercase suffixes](/proposals/p2015.md#uppercase-suffixes)
-   [Additional bit sizes](/proposals/p2015.md#additional-bit-sizes)
```

File: docs/design/expressions/README.md (hunk 8 above)
Comment (josh11b) [inline docs/design/expressions/README.md]:
```suggestion
These are referred to as
[numeric type literals](literals.md#numeric-type-literals).
```

File: docs/design/expressions/README.md [hunk 15]
```diff
@@ -298,6 +299,15 @@ fn Run(args: Span(StringView)) {
 
 `if` expressions are analogous to `?:` ternary expressions in C and C++.
 
+## Numeric type literals expressions
+
+Carbon's syntax provides a simple way to represent different types of integers
+and floating-point numbers. Each type is identified with a keyword-like syntax,
+prefixed with either `i`, `u`, or `f` followed by a multiple of 8, representing
+the size of the data type.
```
Comment (josh11b) [inline docs/design/expressions/README.md]:
```suggestion
the size in bits of the data type.
```

File: docs/design/README.md [hunk 16]
```diff
@@ -469,6 +467,7 @@ may be limited to integers of at most 128 bits due to LLVM limitations.
 
 > References:
 >
+> -   [Numeric type literal expression](expressions/literals.md)
```
Comment (josh11b) [inline docs/design/README.md]:
```suggestion
> -   [Numeric type literal expressions](expressions/literals.md#numeric-type-literals)
```

File: docs/design/README.md [hunk 17]
```diff
@@ -534,6 +533,7 @@ number.
...
```
Comment (josh11b) [inline docs/design/README.md]:
```suggestion
> -   [Numeric type literal expressions](expressions/literals.md#numeric-type-literals)
```

File: docs/design/expressions/README.md [hunk 18]
```diff
@@ -298,6 +299,15 @@ fn Run(args: Span(StringView)) {
 
 `if` expressions are analogous to `?:` ternary expressions in C and C++.
 
+## Numeric type literals expressions
```
Comment (josh11b) [inline docs/design/expressions/README.md]:
```suggestion
## Numeric type literal expressions
```

**Comment by avicorp** [comment]
@zygoloid  @josh11b  Thanks for the Suggestions.

I am struggling to decide what scope should be defined in the overview section of literal expressions. In other languages, I have seen that all literal expressions are grouped together in the same document. However, here, we have split numeric floats and strings under Lexical Conventions. What do you think about what I wrote in the overview?

**Review by josh11b — APPROVED** [review]

**Pull request merged**